File: implementations/go/identity/identity.go
// Package identity implements BlockParty identity derivation (issue #3),
// following protocol/specs/identity.md.
//
// An identity is derived from a user passphrase scoped to a World: the World's
// WalletSalt namespaces the derivation so the same passphrase yields a
// different identity in each World. The signature and KEM keys use
// domain-separated seeds and are therefore independent.
package identity

import (
	"github.com/bpprotocol/blockparty/implementations/go/crypto"
	"github.com/bpprotocol/blockparty/implementations/go/derive"
)

// info strings for the per-purpose HKDF derivations.
const (
	identityInfo = "bpprotocol.org/v1/identity"
	mldsaInfo    = "ml-dsa"
	mlkemInfo    = "mlkem"
)

// emptySalt is the empty (non-nil) HKDF salt the spec specifies for the
// per-key derivations (salt="").
var emptySalt = []byte{}

// Identity is a participant's post-quantum signing key, KEM key, and the
// address that binds them.
type Identity struct {
	Address derive.Address
	MLDSA   crypto.MLDSAKeyPair
	Kyber   crypto.KyberKeyPair
}

// OpenIdentity deterministically derives an identity from a passphrase within a
// World. Recovery is exact: the same (world, passphrase) always reproduces the
// same keys and address, and the same passphrase yields distinct identities
// across Worlds.
func OpenIdentity(w derive.World, passphrase string) Identity {
	worldPassword := crypto.HKDFSHA256([]byte(passphrase), w.WalletSalt, []byte(identityInfo), 32)

	mldsa := crypto.MakeMLDSAPair(crypto.HKDFSHA256(worldPassword, emptySalt, []byte(mldsaInfo), 32))
	kyber := crypto.MakeKyberPair(crypto.HKDFSHA256(worldPassword, emptySalt, []byte(mlkemInfo), 32))

	return Identity{
		Address: derive.BytesToAddress(mldsa.PublicBytes(), kyber.PublicBytes()),
		MLDSA:   mldsa,
		Kyber:   kyber,
	}
}

File: implementations/typescript/packages/sdk/src/identity/identity.ts
import { hkdfSha256, makeKyberPair, makeMldsaPair, type KeyPair } from "../crypto/index.js";
import { bytesToAddress, type World } from "../derive/index.js";
import { utf8ToBytes } from "../bytes.js";

// HKDF info strings for the per-purpose derivations.
const IDENTITY_INFO = "bpprotocol.org/v1/identity";
const MLDSA_INFO = "ml-dsa";
const MLKEM_INFO = "mlkem";

// The empty (non-nil) HKDF salt the spec specifies for the per-key derivations.
const EMPTY_SALT = new Uint8Array(0);

/** A participant's ML-DSA-65 signing key, ML-KEM768 key, and bound address. */
export interface Identity {
  readonly address: string;
  readonly mldsa: KeyPair;
  readonly kyber: KeyPair;
}

/**
 * Deterministically derive an identity from a passphrase within a World. The
 * World's wallet salt namespaces the derivation, so the same passphrase yields
 * a different identity per World; the signature and KEM keys use
 * domain-separated seeds and are independent. Recovery is exact.
 */
export function openIdentity(world: World, passphrase: string): Identity {
  const worldPassword = hkdfSha256(
    utf8ToBytes(passphrase),
    world.walletSalt,
    utf8ToBytes(IDENTITY_INFO),
    32,
  );
  const mldsa = makeMldsaPair(hkdfSha256(worldPassword, EMPTY_SALT, utf8ToBytes(MLDSA_INFO), 32));
  const kyber = makeKyberPair(hkdfSha256(worldPassword, EMPTY_SALT, utf8ToBytes(MLKEM_INFO), 32));
  return { address: bytesToAddress(mldsa.publicKey, kyber.publicKey), mldsa, kyber };
}
